From 6ed083f51a6d626cc8abcbf980db96e6e16f8a67 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Tue, 16 Oct 2018 10:42:07 +0900
Subject: [PATCH 362/909] media: rcar-vin: Add crop capacity get for digital
 interface

commit 4df96a83a15f53be80da15981cc863ec1fd2d5f1 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

Since this driver can not get crop capacity when digital interface,
add this code.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/rcar-vin/rcar-v4l2.c | 44 +++++++++++++++++++++
 1 file changed, 44 insertions(+)

diff --git a/drivers/media/platform/rcar-vin/rcar-v4l2.c b/drivers/media/platform/rcar-vin/rcar-v4l2.c
index b81be2db3905..598d1439ae9f 100644
--- a/drivers/media/platform/rcar-vin/rcar-v4l2.c
+++ b/drivers/media/platform/rcar-vin/rcar-v4l2.c
@@ -407,6 +407,44 @@ static int rvin_enum_fmt_vid_cap(struct file *file, void *priv,
 	return 0;
 }
 
+static int rvin_get_digital_sd_format(struct rvin_dev *vin,
+				      struct v4l2_pix_format *pix)
+{
+	struct v4l2_subdev_format fmt = {
+		.which = V4L2_SUBDEV_FORMAT_ACTIVE,
+	};
+	int ret;
+
+	fmt.pad = vin->digital->source_pad;
+
+	ret = v4l2_subdev_call(vin_to_source(vin), pad, get_fmt, NULL, &fmt);
+	if (ret)
+		goto end;
+
+	switch (fmt.format.field) {
+	case V4L2_FIELD_TOP:
+	case V4L2_FIELD_BOTTOM:
+	case V4L2_FIELD_NONE:
+	case V4L2_FIELD_INTERLACED_TB:
+	case V4L2_FIELD_INTERLACED_BT:
+	case V4L2_FIELD_INTERLACED:
+	case V4L2_FIELD_SEQ_TB:
+	case V4L2_FIELD_SEQ_BT:
+		break;
+	case V4L2_FIELD_ALTERNATE:
+		/* Use VIN hardware to combine the two fields */
+		fmt.format.field = V4L2_FIELD_INTERLACED;
+		break;
+	default:
+		vin->format.field = V4L2_FIELD_NONE;
+		break;
+	}
+
+	v4l2_fill_pix_format(pix, &fmt.format);
+end:
+	return 0;
+}
+
 static int rvin_g_selection(struct file *file, void *fh,
 			    struct v4l2_selection *s)
 {
@@ -423,6 +461,10 @@ static int rvin_g_selection(struct file *file, void *fh,
 			ret = rvin_get_sd_format(vin, &vin->format);
 			if (ret)
 				return ret;
+		} else {
+			ret = rvin_get_digital_sd_format(vin, &vin->format);
+			if (ret)
+				return ret;
 		}
 		s->r.left = s->r.top = 0;
 		s->r.width = vin->source.width;
@@ -925,6 +967,8 @@ static int rvin_initialize_device(struct file *file)
 	if (ret < 0)
 		goto esfmt;
 
+	rvin_get_digital_sd_format(vin, &vin->format);
+
 	v4l2_ctrl_handler_setup(&vin->ctrl_handler);
 
 	return 0;
-- 
2.17.1

